storage/innobase/btr/btr0pcur.cc

```diff
@@ -214,101 +212,68 @@ btr_pcur_copy_stored_position(
+
+  ut_ad(*latch_mode == BTR_SEARCH_PREV || *latch_mode == BTR_MODIFY_PREV);
   const rw_lock_type_t mode=
     rw_lock_type_t(*latch_mode & (RW_X_LATCH | RW_S_LATCH));
 
-  switch (*latch_mode) {
-  default:
-    ut_ad(*latch_mode == BTR_SEARCH_LEAF || *latch_mode == BTR_MODIFY_LEAF);
-    return buf_page_optimistic_get(mode, block, pcur->modify_clock, mtr);
-  case BTR_SEARCH_PREV:
-  case BTR_MODIFY_PREV:
-    page_id_t id{0};
-    uint32_t left_page_no;
-    ulint zip_size;
-    buf_block_t *left_block= nullptr;
-    {
-      transactional_shared_lock_guard<block_lock> g{block->page.lock};
-      if (block->modify_clock != pcur->modify_clock)
-        return false;
-      id= block->page.id();
-      zip_size= block->zip_size();
-      left_page_no= btr_page_get_prev(block->page.frame);
-    }
-
-    if (left_page_no != FIL_NULL)
-    {
-      left_block=
-        buf_page_get_gen(page_id_t(id.space(), left_page_no), zip_size,
-                         mode, nullptr, BUF_GET_POSSIBLY_FREED, mtr);
-
-      if (!left_block);
-      else if (btr_page_get_next(left_block->page.frame) != id.page_no())
-      {
-release_left_block:
-        mtr->release_last_page();
-        return false;
-      }
-      else
-        buf_page_make_young_if_needed(&left_block->page);
-    }
-
-    if (buf_page_optimistic_get(mode, block, pcur->modify_clock, mtr))
-    {
-      if (btr_page_get_prev(block->page.frame) == left_page_no)
-      {
-        /* block was already buffer-fixed while entering the function and
-        buf_page_optimistic_get() buffer-fixes it again. */
-        ut_ad(2 <= block->page.buf_fix_count());
-        *latch_mode= btr_latch_mode(mode);
-        return true;
-      }
-
-      mtr->release_last_page();
-    }
-
-    ut_ad(block->page.buf_fix_count());
-    if (left_block)
-      goto release_left_block;
-    return false;
+  uint64_t modify_clock;
+  uint32_t left_page_no;
+  const page_t *const page= block->page.frame;
+  {
+    transactional_shared_lock_guard<block_lock> g{block->page.lock};
+    modify_clock= block->modify_clock;
+    left_page_no= btr_page_get_prev(page);
   }
-}
 
-/** Structure acts as functor to do the latching of leaf pages.
-It returns true if latching of leaf pages succeeded and false
-otherwise. */
-struct optimistic_latch_leaves
-{
-  btr_pcur_t *const cursor;
-  btr_latch_mode *const latch_mode;
-  mtr_t *const mtr;
+  const auto savepoint= mtr->get_savepoint();
+  mtr->memo_push(block, MTR_MEMO_BUF_FIX);
 
-  bool operator()(buf_block_t *hint) const
+  if (UNIV_UNLIKELY(modify_clock != pcur->modify_clock))
   {
-    return hint &&
-      btr_pcur_optimistic_latch_leaves(hint, cursor, latch_mode, mtr);
+  fail:
+    mtr->rollback_to_savepoint(savepoint);
+    return false;
   }
-};
+
+  buf_block_t *prev= nullptr;
+  if (left_page_no != FIL_NULL &&
+      !(prev= btr_block_get(*pcur->index(), left_page_no, mode, true, mtr)))
+    goto fail;
+
+  mtr->upgrade_buffer_fix(savepoint, mode);
```

Comment:
@mleich1 provided a core dump where we have the `block->page.zip.fix=FREED+1` inside `mtr_t::upgrade_buffer_fix()`. It seems to me that we must relax the assertion inside that function. After acquiring the latch, we must backtrack if the page had been marked as `FREED` in the data file. Callers must check the status of `mtr_t::upgrade_buffer_fix()` and act accordingly.

Reply:
We can simply relax the assertion and check for `block->page.is_freed()` here. Additionally, `buf_page_get_low()` was missing a check of `FREED` after acquiring the page latch.

Reply:
Make sense. Perhaps we should check other callers of upgrade_buffer_fix too.

Reply:
All other callers of `mtr_t::upgrade_buffer_fix()` are in `SPATIAL INDEX` code, specifically in `rtr_latch_leaves()` and `rtr_search_to_nth_level()`. At least in the latter we may invoke `buf_page_get_gen()` with `rw_latch=RW_NO_LATCH`, and we may be accessing a page that is in the process of being read into the buffer pool. `rtr_latch_leaves()` is also being called from `rtr_pcur_getnext_from_path()`, which likewise may invoke `buf_page_get_gen()` with `rw_latch=RW_NO_LATCH`.

It would be tempting to always acquire (and if applicable, release) the page latch in `buf_page_get_low()`. Unfortunately, doing so would obviously result in a deadlock in `btr_latch_prev()`. Outside that function and the page eviction in `PageConverter::operator()`, the only real user of `RW_NO_LATCH` is `SPATIAL INDEX`. I will try to see how to fix that. Yes, `SPATIAL INDEX` are known to be broken already for other reasons, but we should avoid wilingly breaking it further.

Reply:
The `btr_block_get()` call in `btr_pcur_optimistic_latch_leaves()` would trigger debug assertion failures when the page had been freed in the mean time. The problem is that `btr_block_get()` is using `mode=BUF_GET` instead of `BUF_GET_POSSIBLY_FREED`. In non-debug builds there is no issue, because `nullptr` would be returned, which we are dealing with. I will revise this.

Reply:
3a570f75d4e1d5ad2e55189b60fe77437e71959c replaces the `btr_block_get()`